kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -277,144 +269,124 @@ func (c *loadBalancerController) handleIPPoolUpdate(kvp model.KVPair) {
-				err := c.releaseIP(*attr.AttrPrimary)
-				if err != nil {
-					log.Errorf("Error releasing IPAM for handle %s: %s", *attr.AttrPrimary, err)
+				} else {
+					// Service no longer exists, leak confirmed. Release all IPs allocated with the specific handle
+					log.Infof("Service not found for handle: %s. Releasing unused IPs", *attr.AttrPrimary)
+					err := c.releaseIP(*attr.AttrPrimary)
+					if err != nil {
+						log.Errorf("Error releasing IPAM for handle %s: %s", *attr.AttrPrimary, err)
+					}
 				}
 			}
 		}
 	}
 
 	// Check that all services have assigned IPs as requested, skip if there are no ippools
 	if len(c.ipPools) != 0 {
-		for _, key := range c.lbResourceCache.ListKeys() {
-			obj, empty := c.lbResourceCache.Get(key)
-			if !empty {
-				service := obj.(*v1.Service)
-				if isCalicoManagedLoadBalancer(service, c.cfg.AssignIPs) {
-					if service.Status.LoadBalancer.Ingress == nil ||
-						(len(service.Status.LoadBalancer.Ingress) == 1 &&
-							(*service.Spec.IPFamilyPolicy == v1.IPFamilyPolicyRequireDualStack) || *service.Spec.IPFamilyPolicy == v1.IPFamilyPolicyPreferDualStack) {
-						err := c.assignIP(service)
-						if err != nil {
-							log.Errorf("Error assigning IP to service %s: %s", service.Name, err)
-						}
-					}
+		for _, svc := range services {
+			if svc.Status.LoadBalancer.Ingress == nil ||
+				(len(svc.Status.LoadBalancer.Ingress) == 1 &&
+					(*svc.Spec.IPFamilyPolicy == v1.IPFamilyPolicyRequireDualStack) || *svc.Spec.IPFamilyPolicy == v1.IPFamilyPolicyPreferDualStack) {
+				err = c.assignIP(&svc)
+				if err != nil {
+					log.WithError(err).Errorf("Error assigning IP to svc: %s", svc.Name)
 				}
 			}
 		}
 	}
-}
-
-func (c *loadBalancerController) runWorker() {
-	for c.processNextItem() {
-	}
-}
-
-func (c *loadBalancerController) processNextItem() bool {
-	workqueue := c.lbResourceCache.GetQueue()
-	key, quit := workqueue.Get()
-	if quit {
-		return false
-	}
-
-	err := c.syncToDatastore(key.(string))
-	c.handleErr(err, key.(string))
 
-	workqueue.Done(key)
-	return true
-}
-
-// syncToDatastore processes the next item in queue
-func (c *loadBalancerController) syncToDatastore(key string) error {
-	obj, exists := c.lbResourceCache.Get(key)
-
-	if !exists {
-		// Service does not exist release the IPs
-		return c.releaseIP(key)
-	} else {
-		svc := obj.(*v1.Service)
-		// We don't have IP, assign one
-		if svc.Status.LoadBalancer.Ingress == nil {
-			return c.assignIP(svc)
-		} else {
-			// Service was updated
-			return c.checkStatus(svc)
+	if len(c.servicesToUpdate) != 0 {
+		for key, svcObj := range c.servicesToUpdate {
+			switch svcObj.updateType {
+			case serviceUpdateTypeUPDATE:
+				err = c.releaseIP(svcObj.handle)
+				if err != nil {
+					log.WithError(err).Error("Error releasing Service IP")
+					continue
+				}
+				svc, err := c.clientSet.CoreV1().Services(svcObj.service.Namespace).Get(context.Background(), svcObj.service.Name, metav1.GetOptions{})
```

Comment:
We don't want to be using the clientset to Get() services - we have an informer / indexer that we can use to access the latest version of the Service what we know about from the Watch